blocklib/blocks/annotator/annotator.hpp

```diff
@@ -36,22 +36,15 @@ class annotator : virtual public sync_block
 {
 public:
     typedef std::shared_ptr<annotator> sptr;
+    virtual std::vector<tag_t> data() const = 0;
 
-    static sptr make(uint64_t when, size_t itemsize, size_t num_inputs, size_t num_outputs, tag_propagation_policy_t tpp);
+    /**
+     * @brief Set the implementation to CPU and return a shared pointer to the block instance
+     * 
+     * @return std::shared_ptr<annotator> 
+     */
+    static sptr cpu(uint64_t when, size_t itemsize, size_t num_inputs, size_t num_outputs, tag_propagation_policy_t tpp);
```

Comment:
`make_cpu` probably better.

But on a higher level, how do we expect this to be used? I could imagine an automatic choice where the user wants the code to make a best guess. By using different function names that's not possible.

C.f. [`for_each` with execution policy](https://en.cppreference.com/w/cpp/algorithm/for_each). I've not done CUDA/OpenCL, but seems to me that an "execution engine" object could be passed in, which for CPU will not do much (although would it be a place for CPU affinity settings somewhere?

Do we expect different set of parameters to be sent to the different implementations? Different values? Different default values?

When it's a user choice, do we expect the make-function-chooser to be code inside YML? What about direct Python/C++ code?

And which user do we want to be able to change these things around? GRC user (can be solved in YML), API user, or at execution time by someone who got possibly a compiled binary? Especially for the last case it becomes tricky if we want to support a way to funnel such parameters from the user to some switch statement. But it may be a worthy use case?

Especially for generated Python (being portable in its final form) it would be nice if we could run the same Python file with OpenCL on a Raspberry Pi as Cuda on a laptop. We may not want the choice to always be automatic, but I think we can do better in how it's controlled.

Reply:
> But on a higher level, how do we expect this to be used? I could imagine an automatic choice where the user wants the code to make a best guess. By using different function names that's not possible.

This has been an ongoing discussion.  I tried sketching out some class hierarchies that would allow a single block to represent multiple possible implementations that could be swapped (with methods for copying state) at runtime.  This got super complicated.  The ideal situation would be something like:
```cpp
auto b = blocks::xxx:::make(params)
b.cpu() // now b is set up to use its cpu implementation
b.cuda() // now b is set up to use its cuda implementation
```
the compromise we got to in the scheduler working group discussions was basically use the current GR abstract factory pattern. This will allow things that are common across all implementations such as the ports and factory parameters to be consistent, but the underlying implementation is like a separate block (just keep the code in the same folder).  

> C.f. [`for_each` with execution policy](https://en.cppreference.com/w/cpp/algorithm/for_each). I've not done CUDA/OpenCL, but seems to me that an "execution engine" object could be passed in, which for CPU will not do much (although would it be a place for CPU affinity settings somewhere?

Interesting - If I understand correctly where you're going with this it may be similar to what is being done in https://github.com/gnuradio/newsched/blob/master/runtime/include/gnuradio/buffer.hpp#LL153-162
where the `buffer_properties` object gets derived and passed in to include initialization specific to the derived class.

> Do we expect different set of parameters to be sent to the different implementations? Different values? Different default values?

I was hoping to avoid this - if the instantiation is too divergent, then it should probably be a different block.  The idea was to have them look the same at the top level.

> When it's a user choice, do we expect the make-function-chooser to be code inside YML? What about direct Python/C++ code?

For the YAML generation (outside this PR), I was thinking that the YML would specify which implementations are available (cpu, cuda), but the user would manually choose which one to use upon instantiation of the block, and that is static.

> And which user do we want to be able to change these things around? GRC user (can be solved in YML), API user, or at execution time by someone who got possibly a compiled binary? Especially for the last case it becomes tricky if we want to support a way to funnel such parameters from the user to some switch statement. But it may be a worthy use case?

GRC - select the domain that the block will be assigned to (cpu is default)
API - call the desired `make_xxx` function 
binary - I'm not sure, will have to think through.  Haven't so far considered this case

> Especially for generated Python (being portable in its final form) it would be nice if we could run the same Python file with OpenCL on a Raspberry Pi as Cuda on a laptop. We may not want the choice to always be automatic, but I think we can do better in how it's controlled.

That is a step further than I have envisioned this, but worth pursuing.

Reply:
> Interesting - If I understand correctly where you're going with this it may be similar to what is being done in https://github.com/gnuradio/newsched/blob/master/runtime/include/gnuradio/buffer.hpp#LL153-162

Yeah, something like that. Not sure what way would be best, though. Overloading `make` means it can't be decided at runtime. Having the object tell you which should be used means boiler plate switch statement. A map would be nice, but doesn't allow for nice fallback if we want that.

Only thinking about it for a minute:

```c++
return std::map<engine::type, func>{
  {engine::type::CPU,  foo_cpu::make},
  {engine::type::CUDA,  foo_cuda::make},
}[eng.id()](parms here)
```

Or something. Except with a nice wrapper that would allow a better error than a crash if choice is not listed.

>> Do we expect different set of parameters to be sent to the different implementations? Different values? Different default values?
> I was hoping to avoid this 

Yeah, I'm not aware of a reason to complicate things.

> binary - I'm not sure, will have to think through. Haven't so far considered this case

Don't let me derail you by making you assume this is a requirement. It may not be.